merkleTree/routes/infobuRoutes.js

```diff
@@ -0,0 +1,145 @@
+const router = require('express').Router()
+const { MerkleTreePrefix } = require('../lib/MerkleTreePrefix')
+const { SHA256 } = require('crypto-js')
+const { keyNodes } = require('../controllers/keyNodes.controller')
+
+const infoBUsTree = new MerkleTreePrefix([], SHA256, {fillDefaultHash: true})
+
+router.post('/leaves', (req, res) => {
+    const infoBUs = req.body.data
+    const leaves = []
+
+    for (let index = 0; index < infoBUs.length; index++) {
+        const infoBU = infoBUs[index];
+        leaves.push({
+            leaf: SHA256(JSON.stringify(infoBU)).toString(),
+            vote: infoBU.votos_validos.map(candidato => ([candidato.nome, candidato.votos]))
+        })
+    }
+    
+    console.log(leaves)
+    infoBUsTree.addLeaves(leaves)
+    return res.json(req.body.data)
+})
+
+router.get('/', (req, res) => {
+    console.log(infoBUsTree.toString())
+    console.log("tree leaves: ", infoBUsTree.getLeafCount())
+    console.log(infoBUsTree.getHexLayersFlat())
+    res.send(infoBUsTree.getHexLayers())
+})
+
+router.get('/root', (req, res) => {
+    const root = infoBUsTree.getRoot()
+    res.send(root)
+})
+
+router.get('/leaves', (req, res) => {
+    if(req.query.id === undefined){
+        const leaves = infoBUsTree.getHexLeaves()
+        console.log(leaves)
+        res.send(leaves)
+        return
+    }
+
+    const leaf = infoBUsTree.getLeaf(req.query.id)
+    console.log(leaf)
+    res.send({
+      "id": req.query.id,
+      "leaf": leaf.leaf.toString('hex'),
+      "vote": leaf.vote
+    })
+})
+  
+router.get('/leaf/:id', (req, res) => {
+    // Return a leaf with id equals to id and its proof.
+    const leaf = infoBUsTree.getLeaf(req.params.id)
+    const root = infoBUsTree.getRoot()
+    const proof = infoBUsTree.getProof(leaf, req.params.id)
+    console.log(proof)
+    console.log(infoBUsTree.verify(proof, leaf, root))
+    res.send({
+      "id": req.params.id,
+      "leaf": leaf,
+      "proof": proof,
+    })
+})
+
+router.get('/nodeKeys', (req, res) => {
+    const i_inicial = req.query.i_inicial
+    const i_final = req.query.i_final
+    if(!i_inicial || !i_final){
+        res.send("Missing parameters i_inicial and i_final")
+        return
+    }
+
+    const leaves = infoBUsTree.getLayers()[0]    
+    res.send(keyNodes(leaves, parseInt(i_inicial), parseInt(i_final)))
+})
+
+router.post('/proof', (req, res) => {
```

Comment:
Esse método parece ser um método de leitura apenas, correto? Se sim, é mais adequado usar um GET.

Isso tudo é convenção: vc pode usar um post para leitura. Mas isso acaba confundindo a cabeça de quem le o código. Vede, por exemplo https://stackoverflow.com/questions/19637459/rest-api-using-post-instead-of-get

Reply:
É que do jeito que está, esse endpoint recebe folhas da árvore no body da chamada, e retorna a prova dessas folhas. 

Uma alternativa seria receber os índices das folhas por parâmetros da url no GET, mas nesse caso acho que é válido também mudar o endpoint da árvore normal, que está com o método POST também. O que acha?

Reply:
É que do jeito que está, esse endpoint recebe folhas da árvore no body da chamada, e retorna a prova dessas folhas.

Uma alternativa seria receber os índices das folhas por parâmetros da url no GET, mas nesse caso acho que é válido também mudar o endpoint da árvore normal, que está com o método POST também. O que acha?

Reply:
Hum, nesse caso, acho que pode manter... Parece que POST pode ser usado para leitura em alguns casos (exceção)

https://stackoverflow.com/a/47065381